protocol/stream/stream.go

```diff
@@ -95,6 +95,12 @@ type Writer interface {
 	WriteSetEnd() error
 	WriteListBegin(l ListHeader) error
 	WriteListEnd() error
+
+	WriteEnvelopeBegin(eh EnvelopeHeader) error
```

Comment:
nit: Given that the other Header types that are part of the protocol are defined in this interface, and EnvelopeHeader is also directly used by this interface, can we put EnvelopeHeader in this file?